Support 'BaseOutput' and subclasses from 'diffusers' in dynamo

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111978

Extending the workarounds for `transformers` `ModelOutput` to cover `diffusers` `BaseOutput`. Together with https://github.com/huggingface/diffusers/pull/5459 it should unblock export for `diffusers` models.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

BowenBao committed:
Support 'BaseOutput' and subclasses from 'diffusers' in dynamo

[ghstack-poisoned]

BowenBao committed:
Update on "Support 'BaseOutput' and subclasses from 'diffusers' in dynamo"


Extending the workarounds for `transformers` `ModelOutput` to cover `diffusers` `BaseOutput`. Together with https://github.com/huggingface/diffusers/pull/5459 it should unblock export for `diffusers` models.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

BowenBao committed:
Update on "Support 'BaseOutput' and subclasses from 'diffusers' in dynamo"


Extending the workarounds for `transformers` `ModelOutput` to cover `diffusers` `BaseOutput`. Together with https://github.com/huggingface/diffusers/pull/5459 it should unblock export for `diffusers` models.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

BowenBao commented:
@jansel PTAL, thanks!

jansel commented on torch/_dynamo/variables/dicts.py:
```diff
@@ -296,10 +296,29 @@ def call_method(
             return super().call_method(tx, name, args, kwargs)
 
 
+def _is_matching_transformers_cls(cls) -> bool:
+    try:
+        from transformers.file_utils import ModelOutput
+
+        return issubclass(cls, ModelOutput)
+    except ImportError:
+        return False
+
+
+def _is_matching_diffusers_cls(cls) -> bool:
+    try:
+        from diffusers.utils import BaseOutput
+
+        return issubclass(cls, BaseOutput)
+    except ImportError:
+        return False
+
+
```
Can this be done without importing these modules?  This could cause memory bloat for people not using this packages.

BowenBao replied:
I can adjust this function to do string check on cls.__module__ first. However the `_patch_once` function will always try to import the base class. Is that acceptable?

jansel reviewed (changes requested):
Tests failing?

BowenBao commented:
There are some weird segfaults happening in CI, I'm trying to debug. So far I cannot repro on my end, but it seems consistent on CI machines.

BowenBao committed:
Squash experimenting changes; on "Support 'BaseOutput' and subclasses from 'diffusers' in dynamo"


Extending the workarounds for `transformers` `ModelOutput` to cover `diffusers` `BaseOutput`. Together with https://github.com/huggingface/diffusers/pull/5459 it should unblock export for `diffusers` models.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge

BowenBao commented:
@pytorchbot merge

BowenBao committed:
rebase on "Support 'BaseOutput' and subclasses from 'diffusers' in dynamo"


Extending the workarounds for `transformers` `ModelOutput` to cover `diffusers` `BaseOutput`. Together with https://github.com/huggingface/diffusers/pull/5459 it should unblock export for `diffusers` models.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge
